File: legacy-typescript/src/runtime/values.ts
import { Type } from "../semantics/types.js";
import { DateValue } from "../util/dates.js";

export type RuntimeValue = {
  type: Type;
  data: unknown;
};

export type ClassObject = {
  className: string;
  fields: Map<string, RuntimeValue>;
};

export type TextFileHandle = {
  kind: "TextFile";
  mode: "READ" | "WRITE" | "APPEND";
  path: string;
  lines: string[];
  index: number;
  buffer: string[];
  open: boolean;
};

export type RandomFileHandle = {
  kind: "RandomFile";
  path: string;
  position: number;
  buffer: Buffer;
  open: boolean;
};

export function makeValue(type: Type, data: unknown): RuntimeValue {
  return { type, data };
}

export function asDate(value: RuntimeValue): DateValue {
  return value.data as DateValue;
}

File: csharp/Prescribe.Core/Runtime/Values.cs
using Prescribe.Core.Semantics;
using Prescribe.Core.Util;

namespace Prescribe.Core.Runtime;

public sealed record RuntimeValue(TypeSymbol Type, object? Data);

public sealed record ClassObject(string ClassName, Dictionary<string, RuntimeValue> Fields);

public enum TextFileMode
{
    Read,
    Write,
    Append
}

public sealed class TextFileHandle
{
    public required TextFileMode Mode { get; init; }
    public required string Path { get; init; }
    public List<string> Lines { get; init; } = new();
    public int Index { get; set; }
    public List<string> Buffer { get; init; } = new();
    public bool Open { get; set; } = true;
}

public sealed class RandomFileHandle
{
    public required string Path { get; init; }
    public int Position { get; set; } = 1;
    public byte[] Buffer { get; set; } = Array.Empty<byte>();
    public bool Open { get; set; } = true;
}

public static class RuntimeValues
{
    public static RuntimeValue Make(TypeSymbol type, object? data) => new(type, data);

    public static DateValue AsDate(RuntimeValue value) => (DateValue)value.Data!;
}
